### PR TITLE
libtensorflow_2: add TensorFlow 2 as a library

### PR DESCRIPTION
**Pull request by mikesperber**
###### Motivation for this change

libtensorflow is currently TensorFlow 1. Tensorflow 2 is available as a Python library, but not as a C library. I went with the naming convention of the Python TensorFlow 2.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [X] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [X] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/libraries/science/math/tensorflow/bin_2.nix [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+  tfType = if cudaSupport then "gpu" else "cpu";
+
+  system = 
+    if      stdenv.isLinux  then "linux"
+    else if stdenv.isDarwin then "darwin"
+    else unavailable;
+
+  platform =
+    if stdenv.isx86_64 then "x86_64"
+    else unavailable;
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/science/math/tensorflow/bin_2.nix:20]: Why all of this? Isn't meta.platforms enough?

**Comment by mikesperber** [comment]
It's the same way in `bin.nix`. I'd be happy to fix it, but I suggest doing that in a separate commit, and then for both `bin.nix` and `bin_2.nix` in the same way.

@SuperSandro2000 Thanks for looking at this!

**Comment by SuperSandro2000** [comment]
Result of `nixpkgs-review pr 103292` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>libtensorflow_2 (libtensorflow-bin_2)</li>
  </ul>
</details>

**Comment by mikesperber** [comment]
@SuperSandro2000 Sorry for the long delay - for some reason, I did not see the check failure. Fixed it now. Anything I can do to move it along?

File: pkgs/development/libraries/science/math/tensorflow/bin_2.nix [hunk 2]
```diff
@@ -0,0 +1,77 @@
+{ stdenv
+, fetchurl
+, addOpenGLRunpath
+, cudaSupport ? false, symlinkJoin, cudatoolkit, cudnn, nvidia_x11
+}:
+
+with stdenv.lib;
+let
+  unavailable = throw "libtensorflow is not available for this platform!";
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/science/math/tensorflow/bin_2.nix:9]: Please use broken and do not throw here.

File: pkgs/development/libraries/science/math/tensorflow/bin_2.nix [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    if      stdenv.isLinux  then "linux"
+    else if stdenv.isDarwin then "darwin"
+    else unavailable;
+
+  platform =
+    if stdenv.isx86_64 then "x86_64"
+    else unavailable;
+
+  rpath = makeLibraryPath ([stdenv.cc.libc stdenv.cc.cc.lib] ++
+            optionals cudaSupport [ cudatoolkit.out cudatoolkit.lib cudnn nvidia_x11 ]);
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/science/math/tensorflow/bin_2.nix:23]:
```suggestion
  rpath = makeLibraryPath ([stdenv.cc.libc stdenv.cc.cc.lib]
    ++ optionals cudaSupport [ cudatoolkit.out cudatoolkit.lib cudnn nvidia_x11 ]);
```

File: pkgs/development/libraries/science/math/tensorflow/bin_2.nix [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+  packages = import ./binary-hashes_2.nix;
+  packageName = "${tfType}-${system}-${platform}";
+  url = packages.${packageName} or unavailable;
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/science/math/tensorflow/bin_2.nix:27]:
```suggestion
```

File: pkgs/development/libraries/science/math/tensorflow/bin_2.nix [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+      ${optionalString cudaSupport ''
+        addOpenGLRunpath $out/lib/libtensorflow.so $out/lib/libtensorflow_framework.so
+      ''}
+    '';
+
+in stdenv.mkDerivation rec {
+  pname = "libtensorflow";
+  inherit (packages) version;
+
+  src = fetchurl url;
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/science/math/tensorflow/bin_2.nix:47]:
```suggestion
  src = fetchurl packages."${tfType}-${system}-${platform}";
```

File: pkgs/development/libraries/science/math/tensorflow/bin_2.nix [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+    Cflags: -I$out/include/tensorflow
+    EOF
+  '';
+
+  meta = {
+    description = "C API for TensorFlow";
+    homepage = "https://www.tensorflow.org/install/lang_c";
+    license = licenses.asl20;
+    platforms = [ "x86_64-linux" "x86_64-darwin" ];
+    maintainers = with maintainers; [ basvandijk ];
```
[the remark this answers is not in the document]
Comment (mikesperber) [reply pkgs/development/libraries/science/math/tensorflow/bin_2.nix:75]: No, I was hoping he'd review and take it under his wing.

File: pkgs/development/libraries/science/math/tensorflow/bin_2.nix [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+
+  platform =
+    if stdenv.isx86_64 then "x86_64"
+    else unavailable;
+
+  rpath = makeLibraryPath ([stdenv.cc.libc stdenv.cc.cc.lib] ++
+            optionals cudaSupport [ cudatoolkit.out cudatoolkit.lib cudnn nvidia_x11 ]);
+
+  packages = import ./binary-hashes_2.nix;
+  packageName = "${tfType}-${system}-${platform}";
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/science/math/tensorflow/bin_2.nix:26]:
```suggestion
```

File: pkgs/top-level/all-packages.nix [hunk 8]
```diff
@@ -1508,11 +1508,23 @@ in
...
+    cudatoolkit = cudatoolkit_10_0;
+    cudnn = cudnn_cudatoolkit_10_0;
+  };
+
   libtensorflow =
     if python.pkgs.tensorflow ? libtensorflow
     then python.pkgs.tensorflow.libtensorflow
     else libtensorflow-bin;
 
+  libtensorflow_2 =
```
Comment (Mic92) [inline pkgs/top-level/all-packages.nix:1523]: We already dropped python3.pkgs.tensorflow1 in master. You can also just update libtensorflow to _2?

**Comment by Mic92** [comment]
Result of `nixpkgs-review pr 103292` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>libtensorflow-bin</li>
  </ul>
</details>

**Comment by mikesperber** [comment]
@Mic92 Thanks!

**Pull request merged**